### ISSUE TITLE
Suggest `Entry::or_default` for `Entry::or_insert_with(Default::default)`

### ISSUE THREAD

**Issue by Arnavion** [rust-lang/rust-clippy#3812]
For code that calls `.or_insert_with(Default::default)` on a `std::collections::btree_map::Entry` or `std::collections::hash_map::Entry`, clippy should suggest [`std::collections::btree_map::Entry::or_default`](https://doc.rust-lang.org/std/collections/btree_map/enum.Entry.html#method.or_default) and [`std::collections::hash_map::Entry::or_default`](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.or_default)

**Comment by splashofcrimson** [comment]
Hey, I'm a new contributor here - is anyone working on this? If not, may I gave it a shot?

**Comment by Manishearth** [comment]
Go ahead! Let us know if you need help!

**Comment by HarrisonMc555** [comment]
@splashofcrimson I just wanted to let you know that if you're wondering how to get started, I think that the lint that i just wrote is really similar. You can find the code [here](https://github.com/HarrisonMc555/rust-clippy/blob/434b66cf82c2c623baf33fe5e30967ad31ee91af/clippy_lints/src/use_last.rs).

If you don't want to do it anymore, I could probably step in and do it.

**Comment by wookietreiber** [comment]
Without knowing clippy internals, is it possible to issue a warning for any `or_insert_with(MyStruct::default)`, i.e. not just for `Default::default`?

**Comment by llogiq** [comment]
I think the right way is to resolve the path and see if it is a `Default` implementation. Otherwise `MyStruct` could have its own `default` function without implementing `Default` (or worse, while implementing `Default` another way).

**Comment by cauebs** [comment]
Seems to me no one is working on this, so I'll give it a go today!

**Comment by cauebs** [comment]
I understand we don't have to suggest `unwrap_or_default` for `{Option,Result}.unwrap_or(Default::default())` because that is already covered by [or_fun_call](https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call), correct?

**Comment by ebroto** [comment]
>I understand we don't have to suggest unwrap_or_default for {Option | Result}.unwrap_or(Default::default()) because that is already covered by or_fun_call, correct?

Yes. FYI you can test this quickly in the playground, Clippy can be executed there.

This would not be linted though:
```rust
Some(42).unwrap_or_else(Default::default);
```
This could be improved to suggest `unwrap_or_default`, but may be the subject of another issue.

And about the maps
```rust
map.entry("42").or_insert(Default::default());
```
Is linted, but recommends `or_insert_with(Default::default)`, which is the subject of this issue. I think it's OK if the suggestion of one lint triggers another lint (as long as it's not circular xD).

This lint should probably be implemented in the `methods` module.
